### PR TITLE
Opam-specific functionality for Obuilder spec files

### PR DESCRIPTION
**Pull request by benmandrew**
OCaml-CI and opam-repo-ci share a lot of platform-specific code for generating OBuilder spec files. This PR is a work-in-progress to extract some of that shared opam-related code out, in a similar manner to `ocaml-dockerfile-opam` in [ocurrent/ocaml-dockerfile](https://github.com/ocurrent/ocaml-dockerfile).

### CODE REVIEW

**Comment by benmandrew** [comment]
This PR contributes towards https://github.com/ocurrent/ocaml-ci/issues/682.

**Comment by benmandrew** [comment]
Side-note:`distro.ml` is mostly copied from `dockerfile-opam` as it felt wrong for an OBuilder spec library to depend on the Dockerfile library! This file accounts for well over half of the new LOC, so it may be useful to extract this out at some point in the future.

**Comment by MisterDA** [comment]
While I understand the reason behind the duplication, it is already cumbersome to keep the distro list in sync with the real-world. Unless we extract it from `dockerfile-opam`, I'm afraid duplication will introduce more problems that it's trying to solve.

File: dune-project [hunk 1]
```diff
@@ -40,3 +40,19 @@
...
+  (fmt (>= 0.8.9))
+  sexplib
+  astring
+  ppx_deriving
+  ppx_deriving_yojson
+  ppx_sexp_conv
+  (ocaml (>= 4.10.0))
+  (ocaml-version (>= 3.6.0))
+  (opam-format (>= 2.1.0))
+  obuilder-spec))
```
Comment (MisterDA) [inline dune-project:58]:
```suggestion
  (obuilder-spec (= :version)))
```

File: test/test_variant.ml [hunk 2]
```diff
@@ -0,0 +1,33 @@
+let test_one s =
+  let t = Ocaml_ci.Variant.of_string s in
+  let s' = Ocaml_ci.Variant.to_string t in
```
Comment (MisterDA) [inline test/test_variant.ml:3]: Now the tests depend on ocaml-ci? They should run standalone, if possible.

**Comment by benmandrew** [comment]
This is no longer planned for the time being.

**Pull request closed without merging**